ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/consumer/CatchupConsumer.java

```diff
@@ -0,0 +1,248 @@
+    final long startMs = clock.millis();
+    Set<TopicPartition> tps = this.topicPartitions.get();
+    while (tps == null) {
+      if (clock.millis() - startMs > WAIT_FOR_ASSIGNMENT_MS) {
+        throw new KsqlException("Timed out waiting for assignment from Latest");
+      }
+      sleepMs.accept(100L);
+      tps = this.topicPartitions.get();
+    }
+    return tps;
+  }
+
+  @Override
+  protected void subscribeOrAssign() {
+    // We should always have a latest consumer running
+    final LatestConsumer latestConsumer = latestConsumerSupplier.get();
+    Preconditions.checkNotNull(latestConsumer,
+        "Latest should always be started before catchup is run");
+    // Assign current offsets from the latest consumer.  If the latest consumer doesn't have its
+    // assignment yet, this might be null, but we then count on the latest updating this consumer
+    // with the assignment when it's received.
+    this.newAssignment(latestConsumer.getAssignment());
+    final Map<Integer, Long> startingOffsets
+        = offsetRange.getEndOffsets().getSparseRepresentation();
+    onNewAssignment(Optional.of(startingOffsets));
+    // Seek to the provided starting offsets
+    for (TopicPartition tp : consumer.assignment()) {
+      if (startingOffsets.containsKey(tp.partition()) && startingOffsets.get(tp.partition()) >= 0) {
+        consumer.seek(tp, startingOffsets.get(tp.partition()));
+      }
+    }
+  }
+
+  protected void afterOfferedRow(final ProcessingQueue processingQueue) {
+    // Since we handle only one request at a time, we can afford to block and wait for the request.
+    while (processingQueue.isAtLimit()) {
+      LOG.info("Sleeping for a bit since queue is full");
+      sleepMs.accept(1000L);
+    }
+  }
+
+  /**
+   * Checks if the catchup consumer is caught up and will even switch over to latest and close this
+   * consumer if finished.
+   * @return If the catchup consumer has switched over.
+   */
+  private boolean checkCaughtUp() {
+    LOG.info("Checking to see if we're caught up");
+    final Supplier<Boolean> isCaughtUp = () -> {
+      final LatestConsumer lc = latestConsumerSupplier.get();
+      if (lc == null) {
+        return false;
+      }
+      final Map<TopicPartition, Long> latestOffsets = lc.getCurrentOffsets();
+      if (latestOffsets.isEmpty()) {
+        return false;
+      }
+      return caughtUp(latestOffsets, currentPositions);
+    };
+
+    final Runnable switchOver = () -> {
+      LOG.info("Switching over from catchup to latest");
+      final LatestConsumer lc = latestConsumerSupplier.get();
+      if (lc == null) {
+        return;
+      }
+      for (final ProcessingQueue processingQueue : processingQueues.values()) {
+        lc.register(processingQueue);
+      }
+      processingQueues.clear();
+      close();
+    };
+    return catchupCoordinator.checkShouldCatchUp(signalledLatest, isCaughtUp, switchOver);
```

Comment:
I haven't looked back to answer that last question, but it does remind me of the way we configure restore consumers in Streams with an "upper bound" so that they will never pass the point they are supposed to catch up to.